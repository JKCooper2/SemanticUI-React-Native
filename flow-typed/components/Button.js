// @flow
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import type { ColorsType } from '~/styles/colors';

export type ButtonPropsType = {
    children?: Node;
    content?: Node;
    onPress?: () => mixed;
    style?: StyleObj;
    primary: boolean;
    secondary: boolean;
    compact: boolean;
    toggle: boolean;
    active: boolean;
    positive: boolean;
    negative: boolean;
    fluid: boolean;
    circular: boolean;
    icon: boolean | string;
    color: ColorsType;
  };