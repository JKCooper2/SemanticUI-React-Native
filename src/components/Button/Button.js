// @flow
import React, { PureComponent } from 'react';
import { TouchableOpacity, Text, View } from 'react-native';
import _ from 'lodash';

import { spacing, colors } from '~/styles';
import Icon from '~/components/Icon';

import type { ButtonPropsType } from '~/types/components/Button';
import type { ColorsType } from '~/types/styles';

const getBackgroundColor = (primary: boolean, secondary: boolean, color: ColorsType, positive: boolean, negative: boolean): string => {
    if (primary) {
        return colors.colors.primary;
    }

    if (secondary) {
        return colors.colors.secondary;
    }

    if (positive) {
        return colors.colors.green;
    }

    if (negative) {
        return colors.colors.red;
    }

    return _.get(colors.colors, [color], colors.button.default);
};

const getTextColor = (primary: boolean, secondary: boolean, color: ColorsType, positive: boolean, negative: boolean): string => {
    if (primary || secondary || color !== colors.button.default || positive || negative) {
        return colors.colors.white;
    }

    return colors.colors.grey;
};

class Button extends PureComponent<ButtonPropsType> {
    static defaultProps = {
        primary: false,
        secondary: false,
        color: colors.button.default,
        compact: false,
        positive: false,
        negative: false,
        fluid: false,
        circular: false,
        icon: false,
        active: false,
        toggle: false,
        content: ''
    };

    render() {
        const { children, onPress, content, primary, secondary, color, compact, positive, negative, fluid, circular, icon } = this.props;

        let selectedChildren = content || children;

        if (_.isString(icon)) {
            let newIcon = ((icon: any): string); // Flow casting
            selectedChildren = <Icon name={newIcon} />;
        }

        const backgroundColor = getBackgroundColor(primary, secondary, color, positive, negative);
        const textColor = getTextColor(primary, secondary, color, positive, negative);
        const buttonPadding = compact || icon ? spacing.padding.tiny : spacing.padding.small;

        const containerAlign = fluid ? 'stretch' : 'flex-start';

        const borderRadius = circular ? 20 : spacing.border.medium;

        return (
            <View style={{ alignItems: containerAlign, marginHorizontal: spacing.padding.tiny }}>
                <TouchableOpacity onPress={onPress} style={{ minWidth: 40, alignItems: 'center', backgroundColor, borderRadius, padding: buttonPadding }}>
                    {_.isString(selectedChildren) ? <Text style={{ color: textColor }}>{selectedChildren}</Text> : selectedChildren}
                </TouchableOpacity>
            </View>
        );
    }
}

export default Button;