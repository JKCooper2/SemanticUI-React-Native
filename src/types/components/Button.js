// @flow
import type { Node } from 'react';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import type { ColorsType } from '../styles';

export type ButtonPropsType = {
    children?: Node;
    content: string;
    onPress?: () => mixed;
    style?: StyleObj;
    primary: boolean;
    secondary: boolean;
    compact: boolean;
    toggle: boolean;
    active: boolean;
    positive: boolean;
    negative: boolean;
    fluid: boolean;
    circular: boolean;
    icon: boolean | string;
    color: ColorsType;
  };