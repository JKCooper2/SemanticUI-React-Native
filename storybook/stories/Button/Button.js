// @flow
import React, { Fragment } from 'react';

import { storiesOf } from '@storybook/react-native';

import Button from '~/components/Button';
import Icon from '~/components/Icon';

const buttonStories = storiesOf('Button', module);

buttonStories.add('Standard', () => (
    <Button>
        Button
    </Button>
));

buttonStories.add('Primary', () => (
    <Button primary>
        Primary
    </Button>
));


buttonStories.add('Secondary', () => (
    <Button secondary>
        Secondary
    </Button>
));

buttonStories.add('Colors', () => (
    <Fragment>
        <Button color="red" content="Red" />
        <Button color="orange" content="Orange" />
        <Button color="yellow" content="Yellow" />
        <Button color="olive" content="Olive" />
        <Button color="green" content="Green" />
        <Button color="teal" content="Teal" />
        <Button color="blue" content="Blue" />
        <Button color="violet" content="Violet" />
        <Button color="purple" content="Purple" />
        <Button color="pink" content="Pink" />
        <Button color="brown" content="Brown" />
        <Button color="grey" content="Grey" />
        <Button color="black" content="Black" />
    </Fragment>
));

buttonStories.add('Compact', () => (
    <Fragment>
        <Button compact content="Compact" />
        <Button content="Standard" />
    </Fragment>
));

buttonStories.add('Positive/Negative', () => (
    <Fragment>
        <Button positive content="Positive" />
        <Button negative content="Negative" />
    </Fragment>
));

buttonStories.add('Fluid', () => (
    <Button fluid content="Fluid" />
));


buttonStories.add('Circular', () => (
    <Button circular content="C" />
));

buttonStories.add('Icon', () => (
    <Fragment>
        <Button icon>
            <Icon name="rocket" />
        </Button>
        <Button icon="rocket" />
    </Fragment>
));

export default buttonStories;