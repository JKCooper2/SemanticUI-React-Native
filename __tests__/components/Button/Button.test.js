// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Button from '~/components/Button';

test('renders correctly', () => {
    const button = shallow(<Button />);
    expect(button).toMatchSnapshot();
});
