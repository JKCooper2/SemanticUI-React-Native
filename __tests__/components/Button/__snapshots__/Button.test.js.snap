// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "marginHorizontal": 5,
    }
  }
>
  <TouchableOpacity
    activeOpacity={0.2}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#e8e8e8",
        "borderRadius": 5,
        "minWidth": 40,
        "padding": 10,
      }
    }
  />
</Component>
`;
